### DISCUSSION TITLE
Is it possible to bulk delete with composite primary keys?

### DISCUSSION THREAD

**Discussion by mikeyhew** [diesel-rs/diesel#4047] [Q&A] [answered]
I have a query I want to write that involves deleting several rows in a table like this:

```sql
create table student_classes (student_id bigint not null, class_id bigint not null, primary key (student_id, class_id));
```

Which can be done with the following sql

```sql

delete from student_classes where (student_id, class_id) in ((1, 5), (2,4), (10, 8))
```

Is there a way to do this in diesel? I tried the following code but it complains that the tuple type `(diesel::sql_types::Bigint, diesel::sql_types::Bigint)` needs to implement the trait `SingleValue`.

```rust
let student_class_ids_to_delete: &[(i64, i64)] = ...;

diesel::delete(student_classes::table)
    .filter((student_id, class_id).eq_any(student_class_ids_to_delete)
    .execute(conn);
```

Currently I'm falling back to deleting the rows one by one, which works but is slower.

**Comment by weiznich** [comment, marked as answer]
It's always hard to answer questions that ask "Is there any way to do this in diesel" in a meaningful way as you can extend diesel in a lot of ways and based on that almost anything is possible, but just answering these questions with yes it's possible is likely not helpful for anyone.

The built-in DSL does not provide this feature, you would need to write your own extension in this case or use `sql_query`. We are also open to accept contributions for adding support for general operations on tuples to diesel as discussed in https://github.com/diesel-rs/diesel/discussions/3723

**Reply by mikeyhew** [reply]
Thanks, that was what I meant. I'll comment on that discussion.
